aip/0166.md

```diff
@@ -0,0 +1,78 @@
+---
+aip:
+  id: 166
+  state: reviewing
+  created: 2020-06-24
+permalink: /166
+redirect_from:
+  - /0166
+---
+
+# Hierarchically-derived resources
+
+Services occasionally have resources where the resource's effective values are
+determined by consulting both configuration on the resource as well as
+configuration elsewhere, such as a parent in the hierarchy, and "flattening"
+them together. This is most common for configuration resources, which could be
+applied against a resource or a container.
+
+In these situations, users need to distinguish between properties applied to
+the resource directly and properties that are inherited.
+
+## Guidance
+
+Services **should** return only field values on the resource when responding to
+`Get` and `List` methods by default.
+
+Additionally, services **may** add a field to `Get` and `List` requests that
+allows the user to opt-in to receiving the effective configuration instead:
```

Comment:
It would be really helpful to give a concrete example of what would be populated in the book when `effective` is true.

Basically I don't quite understand the use case here, and an example would help.

Reply:
It might be useful to not do the book example for this and actually talk about settings objects, which are the primary use case.

Reply:
@sir-buckyball, can you help here? What would be a good example case?

Reply:
This seems different in a couple of subtle ways that it is probably important to tease out. There may be two distinct use cases.

I was considering a single resource (e.g. `BookSettings`) where the values could be directly set or inherited by a settings object attached to an ancestor.

This example is about a _collection of resources_ where the resource can be in the collection by virtue of being there directly, or by being attached somehow to a parent. Unless `labels` was a _field_ on the `projects/123/storageBuckets/abc` resource, which would then be back to the first use case.

I expect that as we have more situations where a centralized API manages a certain piece of metadata for arbitrary resources all over the place, the second use case will come up more often.

Reply:
I believe you are right that there are likely distinct but related cases here.

I didn't illustrate it well in the example I gave, but in GCP today, both resources (along with many others) store labels as fields with no common API to mutate them.
https://cloud.google.com/resource-manager/reference/rest/v1/projects
https://cloud.google.com/storage/docs/json_api/v1/buckets

Storing labels in a centralized API would be another way to represent labels (as a formal resource as opposed to a field). I had used a centralized API in this example for clarity.